### DISCUSSION TITLE
unexpected behaviour of filter with multipe arguments?

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#58] [General]
>> I'm using LASlib the library developed by LAStools to read and write LAS ans LAZ files. Sometime I do no know all the features myself. I'm not sure how these features are working and are poorly documented in LAStools itself. I don't know if filter_and is even useful because -keep_first -keep_class 2 is already an AND operation.

> I just tested it and we can do `-keep_first -keep_class 2 -filter_and`, and also `-keep_first -keep_class 2 -filter_or`. Actually the equivalent of `-keep_first -keep_class 2` is `-keep_first -keep_class 2 -filter_or` by which we get all points which are EITHER first returns OR ground points. By `-keep_first -keep_class 2 -filter_and` we get all points which are first returns AND classified as ground.  

_Originally posted by @wiesehahn in https://github.com/r-lidar/lasR/discussions/46#discussioncomment-9418281_

I just came across the `filter_and` and `filter_or` options again. 
While the statements 
> the equivalent of `-keep_first -keep_class 2` is `-keep_first -keep_class 2 -filter_or` 

seems true,

the statements that
> `-keep_first -keep_class 2 -filter_or` by which we get all points which are EITHER first returns OR ground points

seems not true

With `drop` the behaviour fits my understanding, but with `keep` I think its weird. I am a little confuse now, maybe everything is correct, but maybe not. I cant clearly think about it anymore, but possibly the following examples give some hints on what I thought was going on when using multiple filters.
Some notes by Martin about the behaviour with `drop` (which seems correct anyway) [here](https://groups.google.com/g/lastools/c/0MORD838IsM)

``` r
library(lasR)
f <- system.file("extdata", "Megaplot.laz", package="lidR")


# intended: keep points which are first returns and also keep points which are of classification 2
# actual: keep points which are first returns and of classification 2
f1 <- exec(reader_las(filter = "-keep_first -keep_class 2") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f1@data$Classification))
#>    2 
#> 5032
summary(as.factor(f1@data$ReturnNumber))
#>    1 
#> 5032

# intended: keep points which are first returns and also keep points which are of classification 2
# actual: keep points which are first returns and of classification 2
f2 <- exec(reader_las(filter = "-keep_first -keep_class 2 -filter_or") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f2@data$Classification))
#>    2 
#> 5032
summary(as.factor(f2@data$ReturnNumber))
#>    1 
#> 5032

# intended: keep points which are first returns and of classification 2
# actual: keep points which are first returns and keep points which are classification 2
f3 <- exec(reader_las(filter = "-keep_first -keep_class 2 -filter_and") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f3@data$Classification))
#>     1     2 
#> 50724  7389
summary(as.factor(f3@data$ReturnNumber))
#>     1     2     3     4 
#> 55756  1281   929   147



# intended: drop points which are first returns and drop points which are of classification 2 (drop points which are either first returns or of classification 2)
# actual: same
f4 <- exec(reader_las(filter = "-drop_first -drop_class 2") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f4@data$Classification))
#>     1 
#> 23477
summary(as.factor(f4@data$ReturnNumber))
#>     2     3     4 
#> 20212  3070   195

# intended: drop points which are first returns and drop points which are of classification 2 (drop points which are either first returns or of classification 2)
# actual: same
f5 <- exec(reader_las(filter = "-drop_first -drop_class 2 -filter_or") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f5@data$Classification))
#>     1 
#> 23477
summary(as.factor(f5@data$ReturnNumber))
#>     2     3     4 
#> 20212  3070   195

# intended: drop points which are first returns AND also of classification 2
# actual: same
f6 <- exec(reader_las(filter = "-drop_first -drop_class 2 -filter_and") + write_las(), on = f) |> 
  lidR::readLAS()
summary(as.factor(f6@data$Classification))
#>     1     2 
#> 74201  2357
summary(as.factor(f6@data$ReturnNumber))
#>     1     2     3     4 
#> 50724 21493  3999   342
```

**Comment by wiesehahn** [comment]
Ok, I found some explanation (and also the confusion) 
https://groups.google.com/g/lastools/c/YQ_suFQ-C9c/m/m8jmkUr2AQAJ

**Comment by Jean-Romain** [comment]
Thank you for sharing. Honestly I'm more confused now than before. The boolean logic seems inverted. For simple cases you should see the filters as a pipeline. In `-keep_first -keep_class 2` the first returns pass the first test and are tested for class 2 while non first return are discarded and not even tested for ground. This is effectively an `AND` operation. However seems to be equivalent to `-keep_first -keep_class 2 -filter_or`. From that point this gives me a headache.

**Reply by wiesehahn** [reply]
This is exactly what I was struggling with. But at least now I know how it works, and because it is confusing I might remember :)
